# Ethereum's Relation to Crypto, Stocks, and Twitter
## Cole Hanlon, 2022

Cryptocurrency is a hot topic in 2022. Understanding how pricing is impacted by other crypto, stocks, and Twitter activity is an interesting study. To create a systematic approach this analysis will be broken into several sections. There will be a focus on Ethereum (ETH) which is a large market cap crypto, relating it to other crytpo, stocks, and Twitter activity. The main goal is determining strengths of correlations of ETH price to these three. The finding's goal is to create an ETH trading bot in the future, and use these three in different strengths based on the findings of the analysis. Then, to determine if these findings are unique to ETH and stop a counter-argument that all crypto behaves the same, we will run similar tests on a small cap crypto.

## Disclaimers



*   API keys have been removed, you must enable twarc2, and receive a key from CoinAPI.io
*   All queries to APIs have been tested in external notebook to save space and ensure data is clean and error free when received



Install plotly

In [ ]:
!pip install plotly

Install twarc2 to enable Twitter API requests

In [ ]:
! pip install twarc
! twarc2 configure

Import all needed libraries

In [ ]:
import tweepy
import json
import requests
import pandas as pd
import matplotlib.pyplot as plt
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

# ETH and Crypto

To remain consistent, we will always compare a crypto against Bitcoin (BTC). BTC is the largest most well known cryptocurrency. BTC is talked about on many mainstream media platforms, and the crypto market is condensed down to BTC by these media platforms oftentimes. We will compare ETH and BTC to understand ETH's relationship to the crypto market.

Using CoinAPI to collect crypto pricing data

In [ ]:
url = 'https://rest.coinapi.io/v1/assets'
headers = {'X-CoinAPI-Key' : '<INSERT YOUR KEY HERE>'} #Free API key limited to 100 requets per day

Collect BTC minute pricing data for the last complete 7 days

In [ ]:
btc_url = 'https://rest.coinapi.io/v1/ohlcv/BTC/USD/history?period_id=1MIN&time_start=2022-04-26T00:00:00&time_end=2022-05-03T00:00:00&limit=100000'
btc_response = requests.get(btc_url, headers=headers)

In [ ]:
print(btc_response)

<Response [200]>


Turn JSON into Pandas DF

In [ ]:
btc_price = pd.DataFrame(json.loads(btc_response.text))
btc_price.head()

time_period_start               time_period_end  \
0  2022-04-26T00:00:00.0000000Z  2022-04-26T00:01:00.0000000Z   
1  2022-04-26T00:01:00.0000000Z  2022-04-26T00:02:00.0000000Z   
2  2022-04-26T00:02:00.0000000Z  2022-04-26T00:03:00.0000000Z   
3  2022-04-26T00:03:00.0000000Z  2022-04-26T00:04:00.0000000Z   
4  2022-04-26T00:04:00.0000000Z  2022-04-26T00:05:00.0000000Z   

                      time_open                    time_close  price_open  \
0  2022-04-26T00:00:00.2416720Z  2022-04-26T00:00:59.8872300Z    40440.90   
1  2022-04-26T00:01:00.1970000Z  2022-04-26T00:01:59.9796860Z    40493.40   
2  2022-04-26T00:02:00.2448650Z  2022-04-26T00:02:59.0731330Z    40484.35   
3  2022-04-26T00:03:00.1720000Z  2022-04-26T00:03:59.9565370Z    40500.87   
4  2022-04-26T00:04:00.0930610Z  2022-04-26T00:04:59.7725420Z    40523.48   

   price_high  price_low  price_close  volume_traded  trades_count  
0   40507.002   40419.45     40490.41      51.014105           825  
1   40500.000   40467.88     40481.87       9.894220           322  
2   40520.000   40477.98     40504.65      14.793987           513  
3   40599.900   40488.00     40523.53      65.757646          1171  
4   40526.120   40449.69     40471.19       4.945948           433

In [ ]:
btc_price.tail()

time_period_start               time_period_end  \
10075  2022-05-02T23:55:00.0000000Z  2022-05-02T23:56:00.0000000Z   
10076  2022-05-02T23:56:00.0000000Z  2022-05-02T23:57:00.0000000Z   
10077  2022-05-02T23:57:00.0000000Z  2022-05-02T23:58:00.0000000Z   
10078  2022-05-02T23:58:00.0000000Z  2022-05-02T23:59:00.0000000Z   
10079  2022-05-02T23:59:00.0000000Z  2022-05-03T00:00:00.0000000Z   

                          time_open                    time_close  price_open  \
10075  2022-05-02T23:55:00.5658430Z  2022-05-02T23:55:59.4766960Z    38511.95   
10076  2022-05-02T23:56:00.1305750Z  2022-05-02T23:56:59.8669070Z    38513.56   
10077  2022-05-02T23:57:00.5056290Z  2022-05-02T23:57:59.1962470Z    38495.46   
10078  2022-05-02T23:58:00.1280000Z  2022-05-02T23:58:59.7926300Z    38509.67   
10079  2022-05-02T23:59:00.2860000Z  2022-05-02T23:59:59.6416830Z    38512.16   

       price_high  price_low  price_close  volume_traded  trades_count  
10075    38543.00   38497.09     38511.83       6.317352           334  
10076    38548.00   38488.85     38495.95      12.805620           407  
10077    38548.00   38494.74     38507.28       3.881117           206  
10078    38545.83   38499.85     38503.17      15.321467           430  
10079    38540.00   38494.73     38510.89      13.384059           330

Collect ETH minute pricing data for the last 7 days

In [ ]:
eth_url = 'https://rest.coinapi.io/v1/ohlcv/ETH/USD/history?period_id=1MIN&time_start=2022-04-26T00:00:00&time_end=2022-05-03T00:00:00&limit=100000'
eth_response = requests.get(eth_url, headers=headers)

In [ ]:
print(eth_response)

<Response [200]>


In [ ]:
eth_price = pd.DataFrame(json.loads(eth_response.text))
eth_price.head()

time_period_start               time_period_end  \
0  2022-04-26T00:00:00.0000000Z  2022-04-26T00:01:00.0000000Z   
1  2022-04-26T00:01:00.0000000Z  2022-04-26T00:02:00.0000000Z   
2  2022-04-26T00:02:00.0000000Z  2022-04-26T00:03:00.0000000Z   
3  2022-04-26T00:03:00.0000000Z  2022-04-26T00:04:00.0000000Z   
4  2022-04-26T00:04:00.0000000Z  2022-04-26T00:05:00.0000000Z   

                      time_open                    time_close  price_open  \
0  2022-04-26T00:00:00.0142730Z  2022-04-26T00:00:59.8747450Z     3007.24   
1  2022-04-26T00:01:00.0970000Z  2022-04-26T00:01:59.9396600Z     3011.30   
2  2022-04-26T00:02:00.7459790Z  2022-04-26T00:02:59.7511930Z     3009.48   
3  2022-04-26T00:03:00.0814890Z  2022-04-26T00:03:59.2888280Z     3010.73   
4  2022-04-26T00:04:00.9090000Z  2022-04-26T00:04:58.6343390Z     3014.19   

   price_high  price_low  price_close  volume_traded  trades_count  
0     3011.34   3004.980      3010.19     286.737292           714  
1     3011.30   3008.375      3009.51     156.284657           470  
2     3011.33   3009.010      3010.66     107.161538           508  
3     3018.20   3009.480      3013.69     621.724265           982  
4     3015.41   3010.183      3011.23     102.263220           463

In [ ]:
eth_price.tail()

time_period_start               time_period_end  \
10075  2022-05-02T23:55:00.0000000Z  2022-05-02T23:56:00.0000000Z   
10076  2022-05-02T23:56:00.0000000Z  2022-05-02T23:57:00.0000000Z   
10077  2022-05-02T23:57:00.0000000Z  2022-05-02T23:58:00.0000000Z   
10078  2022-05-02T23:58:00.0000000Z  2022-05-02T23:59:00.0000000Z   
10079  2022-05-02T23:59:00.0000000Z  2022-05-03T00:00:00.0000000Z   

                          time_open                    time_close  price_open  \
10075  2022-05-02T23:55:00.0243370Z  2022-05-02T23:55:59.6087560Z     2857.01   
10076  2022-05-02T23:56:00.1303380Z  2022-05-02T23:56:59.1626290Z     2857.69   
10077  2022-05-02T23:57:00.8140740Z  2022-05-02T23:57:58.9733730Z     2855.19   
10078  2022-05-02T23:58:01.7310000Z  2022-05-02T23:58:59.8169750Z     2857.07   
10079  2022-05-02T23:59:00.3886980Z  2022-05-02T23:59:59.8962060Z     2855.90   

       price_high  price_low  price_close  volume_traded  trades_count  
10075     2858.03    2854.89      2857.60     116.105354           392  
10076     2858.59    2855.18      2855.19      73.715386           397  
10077     2857.65    2854.87      2857.16     122.300991           470  
10078     2858.52    2855.72      2856.01     131.728847           308  
10079     2857.72    2854.67      2856.21     102.255076           346

Create plots of BTC and ETH prices over the last 7 days to observe retrieved data

In [ ]:
fig1 = px.line(
    btc_price, 
    x='time_period_start', 
    y='price_open',
    color_discrete_sequence=['gold'],
    labels={'start': 'Hour', 'hour_count': 'Tweets per Hour'},
    title='#btc'
)

fig2 = px.line(
    eth_price, 
    x='time_period_start', 
    y='price_open',
    color_discrete_sequence=['plum'],
    labels={'start': 'Hour', 'hour_count': 'Tweets per Hour'},
    title='#eth'
)

fig1.update_layout(
    title="BTC Pricing Last 7 Days",
    xaxis_title="Hour",
    yaxis_title="Price (USD)",
)

fig2.update_layout(
    title="ETH Pricing Last 7 Days",
    xaxis_title="Hour",
    yaxis_title="Price (USD)",
)

fig1.show()
fig2.show()

Simplify the data frames

In [ ]:
btc_price.drop(columns=['time_period_start', 'time_open', 'time_close', 'price_high',	'price_low',	'price_close',	'volume_traded','trades_count', 'time_period_end'], inplace=True)
btc_price.reset_index(inplace=True);

In [ ]:
btc_price = btc_price.rename(columns={"index": "min"})

In [ ]:
eth_price.drop(columns=['time_period_start', 'time_open', 'time_close', 'price_high',	'price_low',	'price_close',	'volume_traded','trades_count', 'time_period_end'], inplace=True)
eth_price = eth_price.reset_index()

In [ ]:
eth_price = eth_price.rename(columns={"index": "min"})

Merge data together on the same minute data

In [ ]:
eth_btc_price = btc_price.merge(eth_price, left_on='min', right_on='min')
eth_btc_price = eth_btc_price.rename(columns={'price_open_x' : 'BTC Price', 'price_open_y' : 'ETH Price', 'min' : 'minute'})
eth_btc_price = eth_btc_price.drop(columns='minute')
eth_btc_price

BTC Price  ETH Price
0       40440.90    3007.24
1       40493.40    3011.30
2       40484.35    3009.48
3       40500.87    3010.73
4       40523.48    3014.19
...          ...        ...
10075   38511.95    2857.01
10076   38513.56    2857.69
10077   38495.46    2855.19
10078   38509.67    2857.07
10079   38512.16    2855.90

[10080 rows x 2 columns]

Run a test

In [ ]:
from scipy.stats import pearsonr

In [ ]:
stat, p = pearsonr(eth_btc_price['BTC Price'], eth_btc_price['ETH Price'])

In [ ]:
eth_btc_price.corr()

BTC Price  ETH Price
BTC Price   1.000000   0.970111
ETH Price   0.970111   1.000000

In [ ]:
print('stat=%.3f, p=%.3f' % (stat, p))

stat=0.970, p=0.000


We can see there is a strong and significant correlation between the price of ETH and the price of BTC.

# ETH and Twitter

Use twarc2 to collect #eth tweet counts hourly over the last 7 days

In [ ]:
! twarc2 counts eth --csv --granularity hour > eth-counts.csv

In [ ]:
eth_counts = pd.read_csv('eth-counts.csv', parse_dates=['start', 'end'])
eth_counts = eth_counts.sort_values('start')
eth_counts.head()

start                       end  hour_count
0 2022-04-26 20:37:01+00:00 2022-04-26 21:00:00+00:00       14851
1 2022-04-26 21:00:00+00:00 2022-04-26 22:00:00+00:00       44430
2 2022-04-26 22:00:00+00:00 2022-04-26 23:00:00+00:00       37559
3 2022-04-26 23:00:00+00:00 2022-04-27 00:00:00+00:00       33625
4 2022-04-27 00:00:00+00:00 2022-04-27 01:00:00+00:00       33792

In [ ]:
eth_counts.tail()

start                       end  hour_count
164 2022-05-03 16:00:00+00:00 2022-05-03 17:00:00+00:00       50356
165 2022-05-03 17:00:00+00:00 2022-05-03 18:00:00+00:00       51564
166 2022-05-03 18:00:00+00:00 2022-05-03 19:00:00+00:00       46345
167 2022-05-03 19:00:00+00:00 2022-05-03 20:00:00+00:00       47069
168 2022-05-03 20:00:00+00:00 2022-05-03 20:37:01+00:00       28076

Create a plot to show retrieved data

In [ ]:
datetimes = pd.to_datetime(eth_counts['start'])

after_hours_eth = eth_counts.copy()
after_hours_eth['hour'] = datetimes.dt.time
after_hours_eth = after_hours_eth.sort_values('hour')
after_hours_eth = after_hours_eth.reset_index();
after_hours_eth = after_hours_eth.drop(columns=['index'])
after_hours_eth = after_hours_eth.drop(after_hours_eth.index[98:147])

after_hours_eth = after_hours_eth.sort_values('start')

fig1 = px.scatter(
    after_hours_eth, 
    x='start', 
    y='hour_count',
    color_discrete_sequence=['#19D3F3'],
    labels={'start': 'Hour', 'hour_count': 'Tweets per Hour'},
    title='#btc'
)

fig2 = px.line(
    eth_counts, 
    x='start', 
    y='hour_count',
    color_discrete_sequence=['rgb(102,102,102)'],
    labels={'start': 'Hour', 'hour_count': 'Tweets per Hour'},
    title='#btc'
)

fig3 = go.Figure(data=fig1.data + fig2.data)

fig3.update_layout(
    title="#ETH Tweets In and Out of U.S. Stock Market Trading Hours",
    xaxis_title="Hour",
    yaxis_title="Hourly Tweet Count",
)

fig3.show()

Collect ETH pricing during the same time interval

In [ ]:
eth_url = 'https://rest.coinapi.io/v1/ohlcv/ETH/USD/history?period_id=1HRS&time_start=2022-04-26T20:37:01&time_end=2022-05-03T20:00:00&limit=100000'
eth_response_tweet = requests.get(eth_url, headers=headers)

In [ ]:
print(eth_response_tweet)

<Response [200]>


Create a new ETH price dataframe

In [ ]:
eth_price = pd.DataFrame(json.loads(eth_response_tweet.text))
eth_price.head()

time_period_start               time_period_end  \
0  2022-04-26T21:00:00.0000000Z  2022-04-26T22:00:00.0000000Z   
1  2022-04-26T22:00:00.0000000Z  2022-04-26T23:00:00.0000000Z   
2  2022-04-26T23:00:00.0000000Z  2022-04-27T00:00:00.0000000Z   
3  2022-04-27T00:00:00.0000000Z  2022-04-27T01:00:00.0000000Z   
4  2022-04-27T01:00:00.0000000Z  2022-04-27T02:00:00.0000000Z   

                      time_open                    time_close  price_open  \
0  2022-04-26T21:00:00.1857990Z  2022-04-26T21:59:58.4402360Z     2829.30   
1  2022-04-26T22:00:00.0102820Z  2022-04-26T22:59:59.7768590Z     2841.74   
2  2022-04-26T23:00:00.1740200Z  2022-04-26T23:59:59.9983030Z     2821.37   
3  2022-04-27T00:00:00.0295690Z  2022-04-27T00:59:58.7190240Z     2809.50   
4  2022-04-27T01:00:00.8571710Z  2022-04-27T01:59:59.7286980Z     2815.62   

   price_high  price_low  price_close  volume_traded  trades_count  
0     2850.00    2820.48      2841.06    7223.109532         26116  
1     2847.16    2819.21      2821.38    8456.488699         29851  
2     2825.22    2762.80      2809.26   16717.155523         41583  
3     2826.10    2794.23      2815.43    9143.508526         30902  
4     2838.99    2814.11      2837.69    7730.078919         26945

In [ ]:
eth_price.tail()

time_period_start               time_period_end  \
162  2022-05-03T15:00:00.0000000Z  2022-05-03T16:00:00.0000000Z   
163  2022-05-03T16:00:00.0000000Z  2022-05-03T17:00:00.0000000Z   
164  2022-05-03T17:00:00.0000000Z  2022-05-03T18:00:00.0000000Z   
165  2022-05-03T18:00:00.0000000Z  2022-05-03T19:00:00.0000000Z   
166  2022-05-03T19:00:00.0000000Z  2022-05-03T20:00:00.0000000Z   

                        time_open                    time_close  price_open  \
162  2022-05-03T15:00:00.1770280Z  2022-05-03T15:59:59.5386020Z     2821.09   
163  2022-05-03T16:00:00.0922210Z  2022-05-03T16:59:59.6221380Z     2821.23   
164  2022-05-03T17:00:00.2524210Z  2022-05-03T17:59:59.7715360Z     2806.07   
165  2022-05-03T18:00:00.3572650Z  2022-05-03T18:59:59.9989770Z     2800.53   
166  2022-05-03T19:00:00.0170000Z  2022-05-03T19:59:59.9220000Z     2781.65   

     price_high  price_low  price_close  volume_traded  trades_count  
162     2836.21    2813.83      2821.24   10202.716327         26836  
163     2827.09    2796.90      2805.75   12786.797976         30267  
164     2826.17    2800.88      2800.88    9999.705075         25621  
165     2801.91    2768.84      2783.08   16717.865416         34431  
166     2784.26    2706.74      2772.98   20411.118960         34763

Simplify tweet count data into simple dataframe

In [ ]:
eth_counts = pd.read_csv('eth-counts.csv', parse_dates=['start', 'end'])
eth_counts = eth_counts.sort_values('start')
eth_counts = eth_counts.drop(eth_counts.index[0])
eth_counts = eth_counts.drop(eth_counts.index[167])
eth_counts.reset_index(inplace=True)
eth_counts.drop(columns=['index', 'start', 'end'], inplace=True)
eth_counts.reset_index(inplace=True)
eth_counts.rename(columns={'index' : 'hour'}, inplace=True)

Simplify ETH price dataframe

In [ ]:
eth_price.drop(columns=['time_period_start',	'time_period_end',	'time_open',	'time_close',	'price_open',	'price_high',	'price_low'	,	'volume_traded',	'trades_count'], inplace=True)
eth_price.reset_index(inplace=True)
eth_price.rename(columns={'index' : 'hour'}, inplace=True)

Merge ETH price and Tweet counts into single dataframe on the same hour

In [ ]:
eth_tweet_price = eth_price.merge(eth_counts, left_on='hour', right_on='hour')
eth_tweet_price.rename(columns={'hour_count' : 'Tweets', 'price_close' : 'ETH Price'},inplace=True)
eth_tweet_price.drop(columns='hour',inplace=True)
eth_tweet_price

ETH Price  Tweets
0      2841.06   44430
1      2821.38   37559
2      2809.26   33625
3      2815.43   33792
4      2837.69   36192
..         ...     ...
162    2821.24   49413
163    2805.75   50356
164    2800.88   51564
165    2783.08   46345
166    2772.98   47069

[167 rows x 2 columns]

Run a test

In [ ]:
eth_tweet_price.corr()

ETH Price    Tweets
ETH Price   1.000000  0.081063
Tweets      0.081063  1.000000

We can see a slight correlation to ETH price and Tweet counts during all hours of the day

Extract Tweet counts outside of trading hours

In [ ]:
eth_counts_after_hours = eth_counts.drop(eth_counts.index[98:147])
eth_counts_after_hours.drop(columns=['hour'],inplace=True)
eth_counts_after_hours

hour_count
0         44430
1         37559
2         33625
3         33792
4         36192
..          ...
162       49413
163       50356
164       51564
165       46345
166       47069

[118 rows x 1 columns]

In [ ]:
eth_counts_after_hours.reset_index(inplace=True)
eth_counts_after_hours.rename(columns={'index' : 'hour'},inplace=True)

In [ ]:
eth_price = pd.DataFrame(json.loads(eth_response_tweet.text))

Extract ETH price data outside of trading hours

In [ ]:
datetimes = pd.to_datetime(eth_price['time_period_start'])

after_hours_eth_price = eth_price.copy()
after_hours_eth_price['hour'] = datetimes.dt.time
after_hours_eth_price = after_hours_eth_price.sort_values('hour')
after_hours_eth_price = after_hours_eth_price.reset_index();
after_hours_eth_price = after_hours_eth_price.drop(columns=['index'])
after_hours_eth_price = after_hours_eth_price.drop(after_hours_eth.index[98:147])
after_hours_eth_price = after_hours_eth_price.sort_values('time_period_start')
after_hours_eth_price.reset_index(inplace=True)
after_hours_eth_price.drop(columns=['index'],inplace=True)

Simplify this ETH price dataframe

In [ ]:
after_hours_eth_price.drop(columns=['time_period_start',	'time_period_end',	'time_open',	'time_close',	'price_open',	'price_high',	'price_low'	,	'volume_traded',	'trades_count', 'hour'], inplace=True)
after_hours_eth_price.reset_index(inplace=True)
after_hours_eth_price.rename(columns={'index' : 'hour'}, inplace=True)

Merge after hours ETH pricing and Tweet counts

In [ ]:
after_hours_total = eth_counts_after_hours.merge(after_hours_eth_price, left_on='hour', right_on='hour')
after_hours_total.rename(columns={'hour_count' : 'Tweets', 'price_close' : 'ETH Price'}, inplace=True)
after_hours_total

hour  Tweets  ETH Price
0      0   44430    2841.06
1      1   37559    2809.26
2      2   33625    2815.43
3      3   33792    2837.69
4      4   36192    2838.48
..   ...     ...        ...
93    93   36772    2757.16
94    94   36042    2764.25
95    95   35546    2758.35
96    96   37331    2760.70
97    97   34092    2806.30

[98 rows x 3 columns]

In [ ]:
after_hours_total.drop(columns=['hour'], inplace=True)

Run a test

In [ ]:
after_hours_total.corr()

Tweets  ETH Price
Tweets     1.000000   0.312845
ETH Price  0.312845   1.000000

We see a very large increase in the correlation of ETH pricing to Tweet counts when outside of trading hours

# ETH and Stocks

We should investigate the correlation of ETH pricing with SPY, the top 500 U.S. stocks, and AAPL the largest tech stock. Many say that crypto tracks U.S. tech stocks, so we should investigate this claim.

Install yfinance

In [ ]:
! pip install yfinance

     |████████████████████████████████| 63 kB 1.3 MB/s 
     |████████████████████████████████| 6.4 MB 32.2 MB/s 
  Attempting uninstall: requests
    Found existing installation: requests 2.23.0
    Uninstalling requests-2.23.0:
      Successfully uninstalled requests-2.23.0
  Attempting uninstall: lxml
    Found existing installation: lxml 4.2.6
    Uninstalling lxml-4.2.6:
      Successfully uninstalled lxml-4.2.6
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires requests~=2.23.0, but you have requests 2.27.1 which is incompatible.
datascience 0.10.6 requires folium==0.2.1, but you have folium 0.8.3 which is incompatible.


In [ ]:
import yfinance as yf

Collect yesterday's $SPY minute data

In [ ]:
ticker_data = yf.Ticker('SPY').history(period='1d', interval='1m')
ticker_data = ticker_data.reset_index();
ticker_data.head()

Datetime        Open        High         Low       Close  \
0 2022-05-03 09:30:00-04:00  415.010010  415.010010  414.260010  414.380005   
1 2022-05-03 09:31:00-04:00  414.359985  414.429993  414.339996  414.380005   
2 2022-05-03 09:33:00-04:00  415.029999  415.500000  414.890015  415.160004   
3 2022-05-03 09:34:00-04:00  415.160004  415.711700  415.160004  415.359985   
4 2022-05-03 09:35:00-04:00  415.369995  415.399994  414.701599  414.779999   

    Volume  Dividends  Stock Splits  
0  2269066          0             0  
1    14964          0             0  
2  1498434          0             0  
3   491258          0             0  
4   476530          0             0

In [ ]:
ticker_data.tail()

Datetime        Open        High         Low       Close  \
385 2022-05-03 15:56:00-04:00  416.760010  416.799988  416.410004  416.500000   
386 2022-05-03 15:57:00-04:00  416.480011  416.739990  416.329987  416.489990   
387 2022-05-03 15:58:00-04:00  416.494995  416.630005  416.350006  416.489990   
388 2022-05-03 15:59:00-04:00  416.500000  416.649994  416.119995  416.410004   
389 2022-05-03 16:00:00-04:00  416.380005  416.380005  416.380005  416.380005   

      Volume  Dividends  Stock Splits  
385   335008          0             0  
386   536222          0             0  
387   490245          0             0  
388  2544666          0             0  
389        0          0             0

Create a plot of the retrieved data

In [ ]:
fig1 = px.line(
    ticker_data, 
    x='Datetime', 
    y='Open',
    color_discrete_sequence=['green'],
    labels={'start': 'Hour', 'hour_count': 'Tweets per Hour'},
    title='$SPY Minute Pricing on 5/3/2022'
)

fig1.show()

Simplify $SPY dataframe

In [ ]:
ticker_data.drop(columns=['Datetime',	'Open',	'High',	'Low',	'Volume',	'Dividends',	'Stock Splits',], inplace=True)
ticker_data.reset_index(inplace=True);
ticker_data.rename(columns={"index": "minute"}, inplace=True)

Collect yesterday's trading hours ETH pricing data

In [ ]:
eth_url = 'https://rest.coinapi.io/v1/ohlcv/ETH/USD/history?period_id=1MIN&time_start=2022-05-03T14:30:00&time_end=2022-05-03T21:00:00&limit=100000'
eth_response_trading = requests.get(eth_url, headers=headers)

In [ ]:
print(eth_response_trading)

<Response [200]>


In [ ]:
eth_price_trading = pd.DataFrame(json.loads(eth_response_trading.text))

Simplify the ETH pricing dataframe

In [ ]:
eth_price_trading.drop(columns=['time_period_start', 'time_open', 'time_close', 'price_high',	'price_low',	'price_close',	'volume_traded','trades_count', 'time_period_end'], inplace=True)
eth_price_trading.reset_index(inplace=True)

In [ ]:
eth_price_trading.rename(columns={"index": "minute"}, inplace=True)

Merge ETH and $SPY pricing

In [ ]:
eth_spy_data = ticker_data.merge(eth_price_trading, left_on='minute', right_on='minute')
eth_spy_data.rename(columns={"Close": "SPY Price", "price_open" : "ETH Price"}, inplace=True)
eth_spy_data.drop(columns=['minute'],inplace=True)
eth_spy_data

SPY Price  ETH Price
0    414.380005    2835.12
1    414.380005    2831.83
2    415.160004    2830.50
3    415.359985    2834.10
4    414.779999    2832.01
..          ...        ...
385  416.500000    2766.34
386  416.489990    2767.79
387  416.489990    2768.76
388  416.410004    2766.53
389  416.380005    2769.19

[390 rows x 2 columns]

Run a test

In [ ]:
eth_spy_data.corr()

SPY Price  ETH Price
SPY Price   1.000000  -0.285349
ETH Price  -0.285349   1.000000

We see that there is actually a negative correlation, we will discuss the ideas of results in the conclusion

Create a dataframe of minute volatility for $SPY

In [ ]:
ticker_data = yf.Ticker('SPY').history(period='1d', interval='1m')
ticker_data = ticker_data.reset_index();
ticker_data.drop(columns=['Datetime', 'High', 'Low', 'Volume', 'Dividends', 'Stock Splits'], inplace=True)
ticker_data['Minute Volatility'] = (ticker_data['Close'] - ticker_data['Open']) / ticker_data['Open'] * 100;
ticker_data

Open       Close  Minute Volatility
0    415.010010  414.380005          -0.151805
1    414.359985  414.380005           0.004831
2    415.029999  415.160004           0.031324
3    415.160004  415.359985           0.048170
4    415.369995  414.779999          -0.142041
..          ...         ...                ...
385  416.760010  416.500000          -0.062388
386  416.480011  416.489990           0.002396
387  416.494995  416.489990          -0.001202
388  416.500000  416.410004          -0.021608
389  416.380005  416.380005           0.000000

[390 rows x 3 columns]

Create a dataframe of minute volatility for ETH

In [ ]:
eth_price_trading = pd.DataFrame(json.loads(eth_response_trading.text))

eth_price_trading.drop(columns=['time_period_start', 'time_open', 'time_close', 'price_high',	'price_low',	'volume_traded','trades_count', 'time_period_end'], inplace=True)
eth_price_trading['Minute Volatility'] = (eth_price_trading['price_close'] - eth_price_trading['price_open']) / eth_price_trading['price_open'] * 100;
eth_price_trading

price_open  price_close  Minute Volatility
0       2835.12      2832.08          -0.107227
1       2831.83      2830.07          -0.062151
2       2830.50      2832.10           0.056527
3       2834.10      2831.56          -0.089623
4       2832.01      2832.63           0.021893
..          ...          ...                ...
385     2766.34      2767.38           0.037595
386     2767.79      2768.76           0.035046
387     2768.76      2766.52          -0.080903
388     2766.53      2769.49           0.106993
389     2769.19      2784.08           0.537702

[390 rows x 3 columns]

Create a dataframe with both volatilities for a plot

In [ ]:
ticker_data['Ticker'] = 'SPY'
eth_price_trading['Ticker'] = 'ETH'
full_volatility = ticker_data.append(eth_price_trading, ignore_index=True)
full_volatility.head()

Open       Close  Minute Volatility Ticker  price_open  price_close
0  415.010010  414.380005          -0.151805    SPY         NaN          NaN
1  414.359985  414.380005           0.004831    SPY         NaN          NaN
2  415.029999  415.160004           0.031324    SPY         NaN          NaN
3  415.160004  415.359985           0.048170    SPY         NaN          NaN
4  415.369995  414.779999          -0.142041    SPY         NaN          NaN

In [ ]:
full_volatility.tail()

Open  Close  Minute Volatility Ticker  price_open  price_close
775   NaN    NaN           0.037595    ETH     2766.34      2767.38
776   NaN    NaN           0.035046    ETH     2767.79      2768.76
777   NaN    NaN          -0.080903    ETH     2768.76      2766.52
778   NaN    NaN           0.106993    ETH     2766.53      2769.49
779   NaN    NaN           0.537702    ETH     2769.19      2784.08

Create a plot comparing minute volatility of $SPY and ETH

In [ ]:
fig = px.box(full_volatility, x="Ticker", y="Minute Volatility")

fig.update_layout(
    title="Minute Volatility of $SPY and ETH on 5/03/22",
    xaxis_title="Ticker",
    yaxis_title="Minute Change (%)",
)

fig.show()

Run a test

In [ ]:
from scipy.stats import f_oneway

In [ ]:
f_oneway(ticker_data['Minute Volatility'], eth_price_trading['Minute Volatility'])

F_onewayResult(statistic=0.861444714604871, pvalue=0.35362256768682954)

We can see that these volatilities are independent. There is a difference shown by the test and in the plot. ETH shows much more volatility than $SPY. Suggesting their pricing will be different.

Collect $AAPL minute data

In [ ]:
ticker_data = yf.Ticker('AAPL').history(period='1d', interval='1m')
ticker_data.reset_index(inplace=True);

Create a plot to show retrieved data

In [ ]:
fig1 = px.line(
    ticker_data, 
    x='Datetime', 
    y='Open',
    color_discrete_sequence=['grey'],
    labels={'start': 'Hour', 'hour_count': 'Tweets per Hour'},
    title='$AAPL Minute Pricing on 5/3/2022'
)

fig1.show()

Simplify data frames

In [ ]:
ticker_data.drop(columns=['Datetime', 'Open',	'High',	'Low',	'Volume',	'Dividends',	'Stock Splits'], inplace=True)
ticker_data.reset_index(inplace=True);
ticker_data.rename(columns={"index": "minute"}, inplace=True)

In [ ]:
eth_price_trading = pd.DataFrame(json.loads(eth_response_trading.text))

eth_price_trading.drop(columns=['time_period_start', 'time_open', 'time_close', 'price_high',	'price_low',	'price_close',	'volume_traded','trades_count', 'time_period_end'], inplace=True)
eth_price_trading.reset_index(inplace=True)

eth_price_trading.rename(columns={"index": "minute"}, inplace=True)

Merge data on each minute of activity

In [ ]:
eth_aapl_data = ticker_data.merge(eth_price_trading, left_on='minute', right_on='minute')
eth_aapl_data.rename(columns={"Close": "AAPL Price", "price_open" : "ETH Price"}, inplace=True)
eth_aapl_data.drop(columns=['minute'],inplace=True)

Run a test

In [ ]:
eth_aapl_data.corr()

AAPL Price  ETH Price
AAPL Price    1.000000  -0.638836
ETH Price    -0.638836   1.000000

We actually end up with an even more inverse relationship between ETH and $AAPL. This suggests that ETH does not track U.S. tech stocks.

# Look at lower market cap crypto

# DOGE and Crypto

Collect data on same interval of DOGE and BTC

In [ ]:
doge_url = 'https://rest.coinapi.io/v1/ohlcv/DOGE/USD/history?period_id=1MIN&time_start=2022-04-26T00:00:00&time_end=2022-05-03T00:00:00&limit=100000'
doge_response = requests.get(doge_url, headers=headers)

In [ ]:
print(doge_response)

<Response [200]>


In [ ]:
doge_price = pd.DataFrame(json.loads(doge_response.text))

Create a plot to show retrieved DOGE data

In [ ]:
fig1 = px.line(
    doge_price, 
    x='time_period_start', 
    y='price_open',
    color_discrete_sequence=['sandybrown'],
    labels={'start': 'Hour', 'hour_count': 'Tweets per Hour'},
    title='DOGE Pricing Last 7 Days'
)
fig1.show()

Clean up data frames

In [ ]:
doge_price.drop(columns=['time_period_start', 'time_open', 'time_close', 'price_high',	'price_low',	'price_close',	'volume_traded','trades_count', 'time_period_end'], inplace=True)
doge_price = doge_price.reset_index()

In [ ]:
doge_price.rename(columns={'index' : 'min'}, inplace=True)

Combine data

In [ ]:
doge_btc_price = btc_price.merge(doge_price, left_on='min', right_on='min')
doge_btc_price = doge_btc_price.rename(columns={'price_open_x' : 'BTC Price', 'price_open_y' : 'ETH Price', 'min' : 'minute'})
doge_btc_price = doge_btc_price.drop(columns='minute')
doge_btc_price

BTC Price  ETH Price
0       40440.90   0.158100
1       40493.40   0.158600
2       40484.35   0.157700
3       40500.87   0.157340
4       40523.48   0.156970
...          ...        ...
10075   38511.95   0.130770
10076   38513.56   0.130800
10077   38495.46   0.130800
10078   38509.67   0.130802
10079   38512.16   0.130800

[10080 rows x 2 columns]

Run a test

In [ ]:
doge_btc_price.corr()

BTC Price  ETH Price
BTC Price   1.000000   0.780627
ETH Price   0.780627   1.000000

# DOGE and Twitter

Collect #DOGE Tweet counts

In [ ]:
! twarc2 counts doge --csv --granularity hour > doge-counts.csv

Read in and clean up

In [ ]:
doge_counts = pd.read_csv('doge-counts.csv', parse_dates=['start', 'end'])
doge_counts = doge_counts.sort_values('start')

doge_counts = doge_counts.drop(doge_counts.index[0])
doge_counts = doge_counts.drop(doge_counts.index[167])
doge_counts.reset_index(inplace=True)
doge_counts.drop(columns=['index', 'start', 'end'], inplace=True)
doge_counts.reset_index(inplace=True)
doge_counts.rename(columns={'index' : 'hour', 'hour_count' : 'Tweets'}, inplace=True)

Collect DOGE pricing on same interval

In [ ]:
doge_url = 'https://rest.coinapi.io/v1/ohlcv/DOGE/USD/history?period_id=1HRS&time_start=2022-04-27T18:27:12&time_end=2022-05-04T18:00:00&limit=100000'
doge_response_tweet = requests.get(doge_url, headers=headers)

In [ ]:
print(doge_response_tweet)

<Response [200]>


In [ ]:
doge_price = pd.DataFrame(json.loads(doge_response_tweet.text))

Clean up DOGE pricing data

In [ ]:
doge_price.drop(columns=['time_period_start',	'time_period_end',	'time_open',	'time_close',	'price_open',	'price_high',	'price_low'	,	'volume_traded',	'trades_count'], inplace=True)
doge_price.reset_index(inplace=True)
doge_price.rename(columns={'index' : 'hour'}, inplace=True)

Combine data

In [ ]:
doge_tweet_price = doge_price.merge(doge_counts, left_on='hour', right_on='hour')
doge_tweet_price.rename(columns={'hour_count' : 'Tweets', 'price_close' : 'DOGE Price'},inplace=True)
doge_tweet_price.drop(columns='hour',inplace=True)
doge_tweet_price

DOGE Price  Tweets
0       0.13855    4338
1       0.13920    4671
2       0.13850    4043
3       0.13820    2831
4       0.14010    2430
..          ...     ...
162     0.13000    4681
163     0.13010    6068
164     0.13100    5439
165     0.13100    4923
166     0.13060    5100

[167 rows x 2 columns]

Run a test

In [ ]:
doge_tweet_price.corr()

DOGE Price    Tweets
DOGE Price    1.000000 -0.013179
Tweets       -0.013179  1.000000

# DOGE and Stocks

In [ ]:
from datetime import date

Collect a full day of trading data

In [ ]:
ticker_data = yf.Ticker('SPY').history(start=date(2022, 5, 3), end=date(2022,5,4), interval='1m')
ticker_data = ticker_data.reset_index();

Clean up this data

In [ ]:
ticker_data.drop(columns=['Datetime',	'Open',	'High',	'Low',	'Volume',	'Dividends',	'Stock Splits',], inplace=True)
ticker_data.reset_index(inplace=True);
ticker_data.rename(columns={"index": "minute"}, inplace=True)

Collect DOGE pricing on same interval

In [ ]:
doge_url = 'https://rest.coinapi.io/v1/ohlcv/DOGE/USD/history?period_id=1MIN&time_start=2022-05-03T14:30:00&time_end=2022-05-03T21:00:00&limit=100000'
doge_response_trading = requests.get(doge_url, headers=headers)

In [ ]:
print(doge_response_trading)

<Response [200]>


In [ ]:
doge_price_trading = pd.DataFrame(json.loads(doge_response_trading.text))

Simplify the DOGE pricing dataframe

In [ ]:
doge_price_trading.drop(columns=['time_period_start', 'time_open', 'time_close', 'price_high',	'price_low',	'price_close',	'volume_traded','trades_count', 'time_period_end'], inplace=True)
doge_price_trading.reset_index(inplace=True)

In [ ]:
doge_price_trading.rename(columns={"index": "minute"}, inplace=True)

Merge DOGE and $SPY pricing

In [ ]:
doge_spy_data = ticker_data.merge(doge_price_trading, left_on='minute', right_on='minute')
doge_spy_data.rename(columns={"Close": "SPY Price", "price_open" : "DOGE Price"}, inplace=True)
doge_spy_data.drop(columns=['minute'],inplace=True)
doge_spy_data

SPY Price  DOGE Price
0    414.380005      0.1302
1    414.070007      0.1302
2    415.220001      0.1303
3    415.160004      0.1304
4    415.359985      0.1304
..          ...         ...
385  416.769989      0.1276
386  416.500000      0.1277
387  416.489990      0.1277
388  416.489990      0.1277
389  416.410004      0.1278

[390 rows x 2 columns]

Run a test

In [ ]:
doge_spy_data.corr()

SPY Price  DOGE Price
SPY Price    1.000000   -0.295215
DOGE Price  -0.295215    1.000000

In [ ]:
#%%shell
#jupyter nbconvert --to html /content/Crypto_Stocks_and_Twitter.ipynb

# Discussion

Overview

The central argument and question at hand for this analysis is what does ETH pricing most strongly correlate to, crypto, stocks, or Twitter activity. The focus is to gain insight on how to build a minute by minute ETH trading bot. A trading bot has a singular goal, to profit and make the most money off of trades. Due to crypto being so volatile and unregulated, it can be very easy to make many trades and receive high percentage gains. A well established crypto trading bot is likely to make lots of profits. The difficulty is the lack of understanding of the crypto market. The stock market has many bankers and analysts to determine future market pricing. However, crypto does not yet have this scene. This could be something to take advantage of. 

ETH and Crypto

The largest crypto currency is Bitcoin (BTC). BTC is mentioned as being the leader in setting the crypto market. With the largest market cap it should be the most stable and influence traders in buying or selling other cryptocurrencies. This is why looking at the correlation between ETH and BTC is so crucial. After collecting data for a previous 7 day period, and plotting both data sets. You can easily see that ETH and BTC have almost the same charts. There are dips and rises in nearly the exact same locations and with the same magnitudes. After filtering down the data sets, we can create a Pearson’s correlation coefficient test of ETH price to BTC price. The results of this test are very interesting. There is a coefficient of 0.97 over the last 7 days. This test is significant according to the P-value. Which means we can nearly perfectly track ETH to BTC. This is extremely important when it comes to creating an ETH trading bot. As we can look towards BTC data to make estimations of future ETH price changes. With such a strong correlation this should prove to be a successful pricing tool.

ETH and Twitter

Twitter is the most popular internet social media platform when it comes to sharing information quickly. There are many pages that allow users who don’t follow you to see your tweets. Hashtags are the most common way of pooling activity together. Users will create a Tweet containing ‘#ETH’, with some other information. When it comes to a trading bot, it would run on minute by minute data. Therefore, we can’t do any highly computational natural language processing. The quickest way of collecting Twitter activity is by counting the number of Tweets with a hashtag during a certain time. To see how Twitter activity changes, we first look at 7 day data. The graph created shows many falls and rises in activity. Interestingly, by overlaying points where the U.S. stock market is closed, there are rapid increases in Twitter activity during the pre-market, and rapid down-fall in the post market. Throughout the trading day Twitter activity does not change much. Therefore, when looking for a correlation we should investigate total day correlations, and after hours correlations. During total days, there is a small correlation between ETH price and the number of Tweets, with a value of 0.08. When running the same test on afterhours data, the correlation increases quite substantially to 0.31. This is a very interesting point and would be crucial to implement in an ETH trading bot. After hours Twitter activity is a much more valid source of ETH pricing data than with trading hours. The bot could dynamically increase and decrease the strength at which it values Twitter activity based on the current time of day. 

ETH and Stocks

The U.S. stock market is by far the most important and looked at the economic market. The stock market determines how the economy is doing, and thus, should have an impact on other markets relating to money. However, each crypto is supposed to be their own currency. This would mean eventually there should not be much of a relationship. Some suggest that BTC is more similar to gold, investors will move money into BTC when the general stock market is down. Others suggest that the crypto market has just become similar to U.S. tech stocks, and crypto prices will follow that category of stock. After collecting data from yfinance, the first stock investigated is SPY. This is the S&P 500 index fund, which is a collection of the top 500 stocks, and is a great overall market indicator. When running Pearson’s correlation coefficient there is a value of -0.28. This suggests a negative relationship between SPY and ETH. This is very interesting, however, it suggests one point of investors, moving money into crypto when the stock market loses value. Another way to investigate a relationship is by volatility. We know that crypto has much higher volatility than the stock market. This is due to being less overall value, and ease of trading. By creating a dataset of minute volatilities during a trading day, a bar and whisker plot is a great visual. We can see that SPY has overall less minute volatility than ETH. ETH also includes outliers which are much more spread apart than the outliers of SPY. This suggests that ETH pricing can fluctuate at a much greater value than SPY. An f-test does not show significance. However, we can still observe this difference, and understand that pricing ETH can be much more challenging than SPY. To investigate the claims that ETH and other crypto simply track the stock market we can investigate the largest tech stock, $AAPL. After running a Pearson’s correlation coefficient test, there is a value of -0.63. This is not hopeful for those who believe crypto follows tech stocks. The idea of relating ETH to U.S. stocks is very difficult. However, we have shown that there is actually a difference, and perhaps we can see dips in stocks, and relate this to an increase in crypto. 

Small Cap Crypto

The biggest counterargument to my idea of creating a specific ETH trading bot, is that ETH is simply like any other crypto currency. Many will believe that the crypto market is all just one, and there is no reason to spend time diving into each individual crypto to build a specific bot. To be sure that this is not a waste of time, we should investigate another crypto. A well spoken about lower market cap crypto is Dogecoin (DOGE). DOGE was very popular amongst Twitter users when it increased significantly in value in just days. Unfortunately, the Twitter API is limited to the previous 7 days. We cannot investigate this time period, however, I would expect an extreme correlation between DOGE price and Twitter activity. We investigate the same three relationships with DOGE. Firstly, looking at DOGE and its relationship to BTC. After running the same Pearson’s correlation coefficient, we see a value of 0.78. This is very strong, however, we can already see that ETH values the price of BTC much more. An ETH bot thus would focus more on BTC, than a DOGE bot would focus on BTC. DOGE does not seem to have any relationship with Twitter activity in the last 7 days. Perhaps, this could change throughout history, and thus a bot should work to figure out if factors begin contributing in stronger ways. DOGE shows a similar negative correlation to SPY. Suggesting similar findings of ETH. In conclusion, a DOGE bot must be different than an ETH bot, and therefore, a bot must be created specific to each crypto.

Conclusion

The analysis is extremely important when looking at ETH in 2022, and understanding how we can predict the price of ETH. Other crypto, Twitter activity, and stocks all seem to play a role. To answer the original argument, a bot would need to focus on other data in this order, BTC price changes, afterhours Twitter activity, the inverse of U.S. stock market data, and trading hours Twitter activity. The values at which an algorithm would weigh these factors could be determined through an analysis over a greater period of time. To counter that ETH behaves the same as other crypto, we investigated DOGE, and showed that there is a difference in pricing impacts. In general anything dealing with financials is very hard to predict, there could be many other impacts on pricing, and compound relationships that were not investigated. Therefore, this is in no way financial advice.
